src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -355,48 +374,102 @@ float _GGXSmith_G1_wo_numerator(in float a2, in float NdotX)
+float irr_glsl_ggx_smith_height_correlated_approx_wo_numerator(in float a, in float NdotL, in float NdotV)
 {
     return 0.5 / mix(2.0*NdotL*NdotV, NdotL+NdotV, a);
 }
 
 //Taken from https://google.github.io/filament/Filament.md.html#materialsystem/anisotropicmodel
-float GGXSmithHeightCorrelated_aniso_wo_numerator(in float at, in float ab, in float TdotL, in float TdotV, in float BdotL, in float BdotV, in float NdotL, in float NdotV)
+float irr_glsl_ggx_smith_height_correlated_aniso_wo_numerator(in float at, in float ab, in float TdotL, in float TdotV, in float BdotL, in float BdotV, in float NdotL, in float NdotV)
 {
     float Vterm = NdotL * length(vec3(at*TdotV, ab*BdotV, NdotV));
     float Lterm = NdotV * length(vec3(at*TdotL, ab*BdotL, NdotL));
     return 0.5 / (Vterm + Lterm);
 }
 
+#endif
+)";
+    }
+    static std::string getGGX_cos_eval(const std::string&)
+    {
+        return
+R"(#ifndef _IRR_BSDF_BRDF_SPECULAR_GGX_INCLUDED_
+#define _IRR_BSDF_BRDF_SPECULAR_GGX_INCLUDED_
+
+#include <irr/builtin/glsl/bsdf/common.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/ndf/ggx.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/geom/ggx.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/fresnel/fresnel.glsl>
+
+vec3 irr_glsl_ggx_height_correlated_aniso_cos_eval(in irr_glsl_BSDFAnisotropicParams params, in mat2x3 ior2, in float a2, in vec2 atb, in float aniso)
+{
+    float g = irr_glsl_ggx_smith_height_correlated_aniso_wo_numerator(atb.x, atb.y, params.TdotL, params.TdotV, params.BdtoL, params.BdotV, params.isotropic.NdotL, params.isotropic.NdotV);
+    float ndf = irr_glsl_ggx_burley_aniso(aniso, a2, params.TdotH, params.BdotH, params.isotropic.NdotH);
+    vec3 fr = irr_glsl_fresnel_conductor(ior2[0], ior2[1], params.isotropic.VdotH);
+
+    return params.isotropic.NdotL * g*ndf*fr;
+}
+vec3 irr_glsl_ggx_height_correlated_cos_eval(in irr_glsl_BSDFIsotropicParams params, in mat2x3 ior2, in float a2)
+{
+    float g = irr_glsl_ggx_smith_height_correlated_wo_numerator(a2, params.NdotL, params.NdotV);
+    float ndf = irr_glsl_ggx_trowbridge_reitz(a2, params.NdotH*params.NdotH);
+    vec3 fr = irr_glsl_fresnel_conductor(ior2[0], ior2[1], params.VdotH);
+
+    return params.NdotL * g*ndf*fr;
+}
+
+#endif
+)";
+    }
+    static std::string getBeckmannSmith(const std::string&)
+    {
+        return
+            R"(#ifndef _IRR_BSDF_BRDF_SPECULAR_BECKMANN_SMITH_INCLUDED_
+#define _IRR_BSDF_BRDF_SPECULAR_BECKMANN_SMITH_INCLUDED_
+
+#include <irr/builtin/glsl/bsdf/common.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/ndf/beckmann.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/geom/ggx.glsl>
+#include <irr/builtin/glsl/bsdf/brdf/specular/fresnel/fresnel.glsl>
+
+vec3 irr_glsl_beckmann_smith_height_correlated_cos_eval(in irr_glsl_BSDFIsotropicParams params, in mat2x3 ior2, in float a2)
+{
+    float g = irr_glsl_ggx_smith_height_correlated_wo_numerator(a2, params.NdotL, params.NdotV);
```

Comment:
does smith height correlated that we use with GGX make sense with beckmann?

Reply:
erm no idea actually, how do i check?

Reply:
https://www.google.com/search?sxsrf=ACYBGNRxf-fDA8g4JKG157ge5izXRbUgFA%3A1581049323375&ei=6-U8XvyvFsHdwALT1rWABg&q=smith+masking+shadowing+beckmann&oq=smith+masking+shadowing+beckmann&gs_l=psy-ab.3...6643.8324..8562...0.2..0.167.1222.0j9......0....1..gws-wiz.......0i71j0i22i30j33i160.brAsgFOVqUw&ved=0ahUKEwi80aray77nAhXBLlAKHVNrDWAQ4dUDCAs&uact=5

Reply:
http://jcgt.org/published/0003/02/03/paper.pdf

Reply:
Skip straight to section 6

Reply:
where you can find 

Height Correlated:
https://www.researchgate.net/profile/Guy_Potvin/publication/7467199_Detailed_analytical_approach_to_the_Gaussian_surface_bidirectional_reflectance_distribution_function_specular_component_applied_to_the_sea_surface/links/554fdf6b08ae956a5d246a1e/Detailed-analytical-approach-to-the-Gaussian-surface-bidirectional-reflectance-distribution-function-specular-component-applied-to-the-sea-surface.pdf

Heitz Weird Shit:
https://hal.archives-ouvertes.fr/hal-00840796/document

Reply:
Basically smith height correlated function always has the same form
`1/(1+\biglambda(\omega_o)+\biglambda(\omega_i))`

Its just that for a different NDF, the `\biglambda` is different (first paped pdf I linked quotes Walter 2007 for a rational approximation of this function for Beckmann)

P.S. We might want to upgrade to Height-Direction-Correlated Masking and Shadowing at some point

Reply:
`a==NdotSomething*inversesqrt(1-NdotSomething*NdotSomething)/(alpha)` instead of `a==1/(alpha*tan(acos(NdotSomething))`

Reply:
put the Walter approximation into a symbolic manipulation package (whatever we used for the fresnel correction last time) and let it simplify it for an equation in terms of `alpha` and `NdotSomething`